### ISSUE TITLE
[Feature] Ref Parsing & Manipulation Functionality

### ISSUE THREAD

**Issue by paul-cureton** [regclient/regclient#760]
It is often needed to process the data embedded in a reference.  Please consider including functionality to work with references directly.  Here are a few examples:

### regctl
Parse a reference into structured data that other tools or code can use:
```
# Splitting across lines for readability, probably better to actually print on one line
$ regctl reference inspect busybox
> {
  "registry": "docker.io",
  "repository": "library/busybox",
  "tag": "latest",
  "digest": null,
  "name": "library/busybox:latest"
}

$ regctl reference inspect --verbose http://my.test:5000/a/b/c@sha512:<...val>
> {
  "scheme": "http",
  "insecure": true,
  "host": "my.test",
  "port": 5000,
  "registry": "my.test:5000",
  "repositoryParts": ["a", "b", "c"],
  "repository": "a/b/c"
  "tag": "latest"
  "digest": "sha512:<...val>",
  "algorithm": "sha512",
  "encode": "<...val>",
  "tag": null,
  "nameSuffix": "@sha512:<...val>",
  "name": "a/b/c@sha512:<...val>"
}
```

Format references to normalize or manipulate them:
```
$ regctl reference normalize busybox
> docker.io/library/busybox:latest

$ regctl reference normalize busybox --format "myregistry.test/{{ .name }}"
> myregistry.test/library/busybox:latest

$ source="myregistry.test/dev/team/service:latest"
$ format='{{ .registry }}/qa/{{ slice .repositoryParts 1 | join "/" }}{{ .nameSuffix }}'
$ target=$(regctl reference normalize "$source" --format "$format")
$ regctl image copy "$source" "$target"
```

### regsync
More Go templates for `regsync`:
```yaml
sync:
  # Simple copy
  - source: alpine
    target: "myregistry.test/{{ .source.name }}"
    type: image
  # Their prod is our dev
  - source: "vendor.test"
    target: "myregistry.test/{{ .source.repository }}:dev"
    tags:
      allow: ["prod"]
    type: registry
```

### regbot
Expanded lua `ref` for `regbot`:
```yaml
scripts:
  - name: a
    script: |
     source = reference.new("debian")
     target = reference.new("myregistry.test/customNamespace/" .. source.name)
     image.copy(source, target)
  - name: b
    script: |
      source = reference.new("debian")
      if source.digest == nil then
        -- Or something like this; unsure if this functionality exists already
        source.digest = tag.resolveDigest(source)
      end
      if source.algorithm ~= "sha512" then
        log("Not quantum resistant")
      else
        target = reference.new("myregistr.test/customNamespace" .. source.repository .. "@" .. source.digest)
        image.copy(source, target)
      done
```

Thanks 🙂

**Comment by sudo-bmitch** [comment]
Interesting request. Are you aware of related tools providing this functionality today? Is there an underlying issue you are trying to resolve (e.g. is this an XY problem)? And if one of these requests is more useful than the others for you, it may be possible to add a more targeted feature.

Some of the hesitation of exposing too much of this is that regclient uses its own reference parser to enable parsing of references to things like OCI Layouts. And reference parsing isn't an OCI standard yet, making it difficult to ensure this parsing matches that of other parsers. It may be better to push for that standardization from OCI first, and then work down to the various projects implementing it.

**Comment by paul-cureton** [comment]
The only tool I am aware of that does something similar is Kyverno, but this particular piece of functionality is only exposed when running in a full K8s cluster in its policy engine.  Aside from that, I only know of a few Go packages used by Skopeo, Crane, and K8s.

I commonly need to copy images, OCI helm charts, and occasionally other types of artifacts between several private registries.  Some of these registries are owned by external groups that use non-unique image/repo names, heavily rely on floating tags, and regularly change the set of images I need to copy.  I can and do maintain several static mappings & scripts myself, but parsing strings is evidently not a strength of mine, and it isn't as consistent as I would hope for.  I did not realize there was no standard for refs, but that would explain why I couldn't find one 😅.

If I could find a way to extract the `registry`, `repository`, and `tag` and/or `digest` consistently, I should be able to parse the more granular fields myself and generate configs for `regsync` or `regbot`.  So if something like the example `regctl reference inspect busybox` command were available, it would make the rest possible.

If you want to wait until there is an official reference spec, though, I'll understand.  Thanks again 🙂

**Comment by sudo-bmitch** [comment]
> So if something like the example regctl reference inspect busybox command were available, it would make the rest possible.

Got it. Adding a hidden/unsupported command for this would be fairly trivial. I'm also considering if a function could be added to the templating sandbox to parse a ref and expose the fields. 

As a heads up on timing, it might be a week or two before I can get to this.

**Comment by sudo-bmitch** [comment]
This should be resolved by #765 now. If it gives you any trouble, let me know.
